File: View/src/services/FrontendService.ts
/**
 * Frontend Service
 * Abstraction layer for frontend operations and API communication.
 * This class represents the frontend component in the system architecture.
 */

import { searchCars, searchCarsWithFilters, getAllCars, healthCheck } from './api';
import type { Car, SearchResponse, Filters } from './api';

export class FrontendService {
  private apiUrl: string;

  constructor(apiUrl?: string) {
    this.apiUrl = apiUrl || import.meta.env.VITE_API_URL || 'http://localhost:5001';
  }

  /**
   * Perform AI-powered car search using natural language query.
   * 
   * @param query - Natural language description of desired car
   * @param filters - Optional filters to apply
   * @param lastId - Optional cursor for pagination
   * @returns Promise resolving to search results
   */
  async searchCars(
    query: string,
    filters?: Filters,
    lastId?: string | number | null
  ): Promise<SearchResponse> {
    try {
      return await searchCars(query, filters, lastId);
    } catch (error) {
      console.error('FrontendService: Search error:', error);
      throw error;
    }
  }

  /**
   * Perform filter-based car search.
   * 
   * @param filters - Filter criteria
   * @param query - Optional search query
   * @param lastId - Optional cursor for pagination
   * @returns Promise resolving to search results
   */
  async searchCarsWithFilters(
    filters: Filters,
    query?: string,
    lastId?: string | number | null
  ): Promise<SearchResponse> {
    try {
      return await searchCarsWithFilters(filters, query, lastId);
    } catch (error) {
      console.error('FrontendService: Filtered search error:', error);
      throw error;
    }
  }

  /**
   * Get all cars (for testing/debugging).
   * 
   * @param limit - Maximum number of cars to return
   * @returns Promise resolving to search results
   */
  async getAllCars(limit: number = 10): Promise<SearchResponse> {
    try {
      return await getAllCars(limit);
    } catch (error) {
      console.error('FrontendService: Get all cars error:', error);
      throw error;
    }
  }

  /**
   * Health check - test if backend is running.
   * 
   * @returns Promise resolving to health status
   */
  async healthCheck(): Promise<{ status: string; message: string }> {
    try {
      return await healthCheck();
    } catch (error) {
      console.error('FrontendService: Health check error:', error);
      throw error;
    }
  }

  /**
   * Normalize car data from API to match frontend Car interface.
   * 
   * @param car - Raw car data from API
   * @returns Normalized car object
   */
  normalizeCar(car: any): Car {
    // Filter out 'gas' and 'unknown' from fuelType
    let fuelType = car.fuel_type || car.fuelType || null;
    if (fuelType) {
      const lower = fuelType.toLowerCase().trim();
      if (lower === 'gas' || lower === 'unknown' || lower === 'null' || lower === '') {
        fuelType = null;
      }
    }

    return {
      id: car.id?.toString() || Math.random().toString(),
      make: car.make || 'Unknown',
      model: car.model || 'Unknown',
      year: car.year || 0,
      price: car.price || 0,
      mileage: car.mileage || 0,
      fuelType: fuelType || '',
      bodyType: car.body_type || car.bodyType || car.carType || 'Unknown',
      color: car.color || 'Unknown',
      image: car.imageUrl || car.image || car.image_url || car.url || car.listing_url || this.getCarImageUrl(car.make, car.model, car.year),
      location: car.location || car.city || 'Unknown',
      url: car.url || car.listing_url,
      colorHex: car.colorHex || car.color_hex || null,
    };
  }

  /**
   * Generate a car image URL from Unsplash based on car details.
   * 
   * @param make - Car manufacturer
   * @param model - Car model
   * @param year - Car year
   * @returns Image URL string
   */
  private getCarImageUrl(make?: string, model?: string, year?: number): string {
    const searchTerms = [make, model, year ? year.toString() : null]
      .filter(Boolean)
      .join(' ');
    
    // Use Unsplash Source API for dynamic car images
    if (searchTerms) {
      const query = encodeURIComponent(`${searchTerms} car`);
      return `https://source.unsplash.com/600x400/?${query}`;
    }
    
    // Fallback to a generic car image from Unsplash
    return 'https://images.unsplash.com/photo-1492144534655-ae79c964c9d7?w=600&h=400&fit=crop&auto=format';
  }
}

// Export singleton instance for convenience
export const frontendService = new FrontendService();

File: Controller/services/frontend_service.py
"""
Frontend Service
Represents the frontend/UI component in the system architecture.
This class is used for sequence diagrams to show the frontend as an object.
"""

from typing import Dict, List, Optional, Any, Union


class FrontendService:
    """
    Service class representing the Frontend/View component.
    This is a representation of the React frontend for sequence diagram purposes.
    The actual frontend is implemented in TypeScript (View/src/).
    """
    
    def __init__(self, api_url: str = "http://localhost:5001"):
        """
        Initialize FrontendService.
        
        Args:
            api_url: URL of the backend API
        """
        self.api_url = api_url
        self.name = "Frontend (React/Vite)"
    
    def search_cars(
        self, 
        query: str, 
        filters: Optional[Dict] = None,
        last_id: Optional[Union[str, int]] = None
    ) -> Dict[str, Any]:
        """
        Perform AI-powered car search using natural language query.
        This represents the frontend calling the backend API.
        
        Args:
            query: Natural language description of desired car
            filters: Optional filter criteria
            last_id: Optional cursor for pagination
        
        Returns:
            Dictionary with search results
        """
        # This is a representation - actual implementation is in TypeScript
        # In sequence diagrams, this shows: Frontend → Backend
        return {
            "method": "POST",
            "endpoint": f"{self.api_url}/api/search",
            "body": {"query": query, "last_id": last_id}
        }
    
    def search_cars_with_filters(
        self,
        filters: Dict,
        query: Optional[str] = None,
        last_id: Optional[Union[str, int]] = None
    ) -> Dict[str, Any]:
        """
        Perform filter-based car search.
        This represents the frontend calling the backend API.
        
        Args:
            filters: Filter criteria
            query: Optional search query
            last_id: Optional cursor for pagination
        
        Returns:
            Dictionary with search results
        """
        # This is a representation - actual implementation is in TypeScript
        # In sequence diagrams, this shows: Frontend → Backend
        return {
            "method": "POST",
            "endpoint": f"{self.api_url}/api/search/filtered",
            "body": {"filters": filters, "last_id": last_id}
        }
    
    def normalize_car(self, car: Dict) -> Dict:
        """
        Normalize car data from API to match frontend Car interface.
        This represents frontend processing of backend responses.
        
        Args:
            car: Raw car data from backend
        
        Returns:
            Normalized car object
        """
        # This is a representation - actual implementation is in TypeScript
        # In sequence diagrams, this shows: Frontend → Frontend (self-call)
        return {
            "id": str(car.get("id", "")),
            "make": car.get("make", "Unknown"),
            "model": car.get("model", "Unknown"),
            "year": car.get("year", 0),
            "price": car.get("price", 0),
            "mileage": car.get("mileage", 0),
            "fuelType": car.get("fuel_type") or car.get("fuelType", ""),
            "bodyType": car.get("body_type") or car.get("bodyType", "Unknown"),
            "color": car.get("color", "Unknown"),
            "image": car.get("imageUrl") or car.get("image", ""),
            "location": car.get("location", "Unknown"),
            "url": car.get("url") or car.get("listing_url"),
            "colorHex": car.get("colorHex") or car.get("color_hex")
        }
    
    def render_car_cards(self, cars: List[Dict]) -> None:
        """
        Render car cards in the user interface.
        This represents the frontend displaying results to the user.
        
        Args:
            cars: List of normalized car objects
        """
        # This is a representation - actual implementation is in TypeScript
        # In sequence diagrams, this shows: Frontend displays results (END - no return arrow)
        pass
    
    def get_name(self) -> str:
        """
        Get the name/identifier of this service.
        
        Returns:
            Service name
        """
        return self.name

